Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,198 @@
+        self.timeInterval = interval
+        if let systemMetricsType = systemMetricsType {
+            self.systemMetricsType = systemMetricsType
+        } else {
+            #if os(Linux)
+            self.systemMetricsType = LinuxSystemMetrics.self
+            #else
+            self.systemMetricsType = NOOPSystemMetrics.self
+            #endif
+        }
+        self.labels = labels
+        self.processId = Int(ProcessInfo.processInfo.processIdentifier)
+
+        self.task = DispatchWorkItem(qos: .background, block: {
+            let metrics = self.systemMetricsType.init(pid: self.processId)
+            if let vmem = metrics.virtualMemoryBytes { Gauge(label: self.labels.label(for: \.virtualMemoryBytes)).record(vmem) }
+            if let rss = metrics.residentMemoryBytes { Gauge(label: self.labels.label(for: \.residentMemoryBytes)).record(rss) }
+            if let start = metrics.startTimeSeconds { Gauge(label: self.labels.label(for: \.startTimeSeconds)).record(start) }
+            if let cpuSeconds = metrics.cpuSeconds { Gauge(label: self.labels.label(for: \.cpuSecondsTotal)).record(cpuSeconds) }
+            if let maxFds = metrics.maxFds { Gauge(label: self.labels.label(for: \.maxFds)).record(maxFds) }
+            if let openFds = metrics.openFds { Gauge(label: self.labels.label(for: \.openFds)).record(openFds) }
+        })
+
+        self.updateSystemMetrics()
+    }
+
+    internal func cancelSystemMetrics() {
+        self.task?.cancel()
+        self.task = nil
+    }
+
+    internal func updateSystemMetrics() {
+        self.queue.asyncAfter(deadline: .now() + self.timeInterval) {
+            guard let task = self.task else { return }
+            task.perform()
+            self.updateSystemMetrics()
+        }
+    }
+}
+
+public struct SystemMetricsLabels {
+    let prefix: String
+    let virtualMemoryBytes: String
+    let residentMemoryBytes: String
+    let startTimeSeconds: String
+    let cpuSecondsTotal: String
+    let maxFds: String
+    let openFds: String
+
+    public init(prefix: String, virtualMemoryBytes: String, residentMemoryBytes: String, startTimeSeconds: String, cpuSecondsTotal: String, maxFds: String, openFds: String) {
+        self.prefix = prefix
+        self.virtualMemoryBytes = virtualMemoryBytes
+        self.residentMemoryBytes = residentMemoryBytes
+        self.startTimeSeconds = startTimeSeconds
+        self.cpuSecondsTotal = cpuSecondsTotal
+        self.maxFds = maxFds
+        self.openFds = openFds
+    }
+
+    func label(for keyPath: KeyPath<SystemMetricsLabels, String>) -> String {
+        return self.prefix + self[keyPath: keyPath]
+    }
+}
+
+public protocol SystemMetrics {
+    init(pid: Int)
+
+    var virtualMemoryBytes: Int? { get }
+    var residentMemoryBytes: Int? { get }
+    var startTimeSeconds: Int? { get }
+    var cpuSeconds: Int? { get }
+    var maxFds: Int? { get }
+    var openFds: Int? { get }
+}
```

Comment:
how about we separate the function and the data? ie make `SystemMetrics` a simple struct that just carried the data (model object if you will) and introduce a protocol `SystemMetricsReader`

```
protocol SystemMetricsReader {
  func read(pid: Int) -> Result<SystemMetrics, Error>
}
```

this way we dont need all these optionals and can simplify the try/catch stuff - imo if it fails than no data better than partial data

Reply:
The reasoning for the optionals here is to always provide as much data as possible to users. For example, on Linux these metrics are retrieved in 3 parts. 1 of those parts could fail, but the other metrics will still be exposed.

Curious what your reasoning is for no data better than partial data. (IMO it's not partial data, just less data)

Reply:
@MrLotU I am interested to understand the failure scenarios. i.e if some of these files you are trying to read does not exist, things must be bad no?